### PR TITLE
Allow replica set to be exposed as a service and add kubectl command tests for replica sets.

### PR DESCRIPTION
**Pull request by madhusudancs**
Only the last commit needs review.

cc @bgrant0607

### CODE REVIEW

**Comment by janetkuo** [comment]
Please include replicasets in `pkg/kubectl/cmd/expose.go` long message and example too.

**Comment by bgrant0607** [comment]
cc @kargakis

**Comment by madhusudancs** [comment]
> Please include replicasets in pkg/kubectl/cmd/expose.go long message and example too.

@janetkuo Done.

**Comment by bgrant0607** [comment]
Note for others: I'm commenting on the last commit.

**Comment by bgrant0607** [comment]
@janetkuo Please move the deployment example from /examples/extensions to /docs/user-guide

**Comment by janetkuo** [comment]
#3024

**Comment by kargakis** [comment]
You also need to handle rs in `PortsForObject` in factory.go for expose.
Can you open issues for rs scaler, rs reaper, and rs autoscaler?

**Comment by madhusudancs** [comment]
> Is kubectl scale implemented using the scale subresource, or does it just update the resource?

It threw me off a bit as well (after reading your comment). Because I remembered commenting out Deployment scaler. I had to go look at the code again to remind myself. We just update the resource - https://github.com/kubernetes/kubernetes/blob/e7a9f309361787679f2210ea4714b79420fc3cba/pkg/kubectl/scale.go#L217

**Comment by kargakis** [comment]
Once the scale subresource is considered ready, I would vote for changing kubectl scale to use the scale endpoints (what already happens with deployments)

**Comment by madhusudancs** [comment]
> You also need to handle rs in PortsForObject in factory.go for expose.

Done.

> Can you open issues for rs scaler, rs reaper, and rs autoscaler?

Scaler and Reaper are already implemented. Filed https://github.com/kubernetes/kubernetes/issues/20986 for kubectl autoscale.

**Comment by madhusudancs** [comment]
@janetkuo @kargakis PTAL.

@bgrant0607 Have addressed all the comments except a README file for replica set docs.

**Comment by kargakis** [comment]
LGTM

File: pkg/kubectl/history.go [hunk 1]
```diff
@@ -67,22 +67,22 @@ func (h *DeploymentHistoryViewer) History(namespace, name string) (HistoryInfo,
 	if err != nil {
 		return historyInfo, fmt.Errorf("failed to retrieve deployment %s: %v", name, err)
 	}
-	_, allOldRCs, err := deploymentutil.GetOldReplicaSets(*deployment, h.c)
+	_, allOldRSs, err := deploymentutil.GetOldReplicaSets(*deployment, h.c)
```
Comment (bgrant0607) [inline pkg/kubectl/history.go]: This fix is unrelated to this PR, but ok.

File: pkg/util/deployment/deployment.go [hunk 2]
```diff
@@ -72,7 +72,7 @@ func GetOldReplicaSetsFromLists(deployment extensions.Deployment, c clientset.In
 	// TODO: Right now we list all replica sets and then filter. We should add an API for this.
 	oldRSs := map[string]extensions.ReplicaSet{}
 	allOldRSs := map[string]extensions.ReplicaSet{}
-	rsList, err := getRcList(namespace, options)
+	rsList, err := getRSList(namespace, options)
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: Tests passed with this broken? Yikes.

**Comment by bgrant0607** [comment]
@madhusudancs If the yaml files are just for the test and not examples for users, then we probably should find a better place for them later.

**Comment by bgrant0607** [comment]
The 3 commits appear in a weird order. Please rebase and squash.

**Comment by bgrant0607** [comment]
Don't change it now, but hack/testdata is probably the right place for test-cmd.sh inputs.

File: pkg/util/deployment/deployment.go (hunk 2 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Just a name. Should be fine.

Comment (madhusudancs) [inline pkg/util/deployment/deployment.go]: `getRcList` (and now `getRSList`) is a local name. That function is passed as an argument to `GetOldReplicaSetsFromLists()` and all the occurrences inside `GetOldReplicaSetsFromLists()` referred to getRcList, so it did not break the tests. But it is still scary. I hope I haven't missed this somewhere else where tests are silently passing.

**Comment by madhusudancs** [comment]
This is next in the queue to be merged and as @bgrant0607 mentioned the commits are in a strange order. I am on a bus and can't be sure when SSH just drops the connection. So removing the LGTM label for now. Will add it back once the PR is rebased and the commits squashed.

**Comment by madhusudancs** [comment]
Everything looks Ok, adding LGTM back.

File: pkg/kubectl/history.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/kubectl/history.go]: This was an artifact of the weird commits ordering, which itself was due to an incorrect rebase. Fixed the rebase and squashed the commits. This has disappeared now.

**Comment by bgrant0607** [comment]
Thanks. LGTM

**Comment by madhusudancs** [comment]
@k8s-bot test this issue: #21045

**Pull request merged**